servers/src/main/java/tachyon/master/file/meta/InodeDirectory.java

```diff
@@ -0,0 +1,213 @@
+
+    return ret;
+  }
+
+  /**
+   * Returns the child with the given inode id.
+   *
+   * @param id The inode id of the child
+   * @return the inode with the given id, or null if there is no child with that id
+   */
+  public synchronized Inode getChild(long id) {
+    return mChildren.getFirstByField(mIdIndex, id);
+  }
+
+  /**
+   * Returns the child with the given name.
+   *
+   * @param name The name of the child
+   * @return the inode with the given name, or null if there is no child with that name
+   */
+  public synchronized Inode getChild(String name) {
+    return mChildren.getFirstByField(mNameIndex, name);
+  }
+
+  /**
+   * Returns the folder's children.
+   *
+   * @return an unmodifiable set of the children inodes.
+   */
+  public synchronized Set<Inode> getChildren() {
+    return ImmutableSet.copyOf(mChildren.iterator());
+  }
+
+  /**
+   * Returns the ids of the children.
+   *
+   * @return the ids of the children
+   */
+  public synchronized Set<Long> getChildrenIds() {
+    Set<Long> ret = new HashSet<Long>(mChildren.size());
+    for (Inode child : mChildren) {
+      ret.add(child.getId());
+    }
+    return ret;
+  }
+
+  /**
+   * Returns the number of children the folder has.
+   *
+   * @return the number of children in the folder.
+   */
+  public synchronized int getNumberOfChildren() {
+    return mChildren.size();
+  }
+
+  /**
+   * Removes the given inode from the folder.
+   *
+   * @param child The Inode to remove
+   * @return true if the inode was removed, false otherwise.
+   */
+  public synchronized boolean removeChild(Inode child) {
+    return mChildren.remove(child);
+  }
+
+  /**
+   * Removes the given child by its name from the folder.
+   *
+   * @param name The name of the Inode to remove.
+   * @return true if the inode was removed, false otherwise.
+   */
+  public synchronized boolean removeChild(String name) {
+    return mChildren.removeByField(mNameIndex, name);
+  }
+
+  @Override
+  public String toString() {
+    StringBuilder sb = new StringBuilder("InodeFolder(");
+    sb.append(super.toString()).append(",").append(getChildren()).append(")");
+    return sb.toString();
+  }
+
+  @Override
+  public synchronized void writeToJournal(JournalOutputStream outputStream)
+      throws IOException {
+    outputStream
+        .writeEntry(new InodeDirectoryEntry(getCreationTimeMs(), getId(), getName(), getParentId(),
+            isPinned(), getLastModificationTimeMs(), getChildrenIds()));
+
+    // Write sub-directories and sub-files via breadth-first, so that during deserialization, it may
+    // be more efficient than depth-first during deserialization due to parent directory's locality.
+    Queue<Inode> children = new LinkedList<Inode>(getChildren());
+    while (!children.isEmpty()) {
+      Inode child = children.poll();
+      child.writeToJournal(outputStream);
+      if (child.isDirectory()) {
+        children.addAll(((InodeDirectory) child).getChildren());
```

Comment:
I restructured a lot of this to prevent this issue. However, I think this is still breadth first, since all children are added at the same time.